### ISSUE TITLE
Map to SPI/BCS SLAVE peripheral SFRs needs to be parsed to “volatile uint32_t *” (pigpio.c, line 7450)

### ISSUE THREAD

**Issue by smeerpate** [joan2937/pigpio#417]

**Comment by guymcswain** [comment]
I don't understand.  Is there a problem you are experiencing with the library?

**Comment by smeerpate** [comment]
Hi,

I was using the RPi BSC slave as an i2c slave and left it run for over a day and a half or so.

I am periodically disabling the peripheral by writing 0 to the I2C bit in the CR register. I’m doing this because i need the i2c slave to NACK a read request from my i2c master and I’m doing so using the control flags in my bsc_xfer_t instance.

After some period, the peripheral wasn’t ACK-ing the i2c master’s read request even when it was supposed to do so. The i2c bit in my bsc_xfer_t instance was effectively set.

Still in this situation, I mmaped the peripheral’s registers using some other code and noticed that the i2c bit there was cleared.

I’m very sorry to have posted this issue and to have bothered you with this. But I’ve changed this in my local fork and didn’t come across this issue any longer. Does this makes sense to you?

Best regards,
Frederic

**Comment by guymcswain** [comment]
No I'm not bothered, I just needed more information so thanks for providing it.

I think I follow what your are doing with the peripheral but sharing a code snippet would be very helpful.

I'm scratching my head over how mapping to volatile resolves the issue.  If there was some kind of coherency issue due to the way the peripherals are mapped then many parts of the library would be falling over.

**Comment by guymcswain** [comment]
I think I found your code here?
https://github.com/smeerpate/SACI2cIotSlave/blob/main/SACRPiIotSlave.c

**Comment by guymcswain** [comment]
Maybe I'm missing it but I don't see where in your code **all** the flags are initialized.

**Comment by smeerpate** [comment]
Hi Guy,
It was my understanding that this was done in the getControlBits function in SACRPiIotSlave.c...

> I think I found your code here?
> https://github.com/smeerpate/SACI2cIotSlave/blob/main/SACRPiIotSlave.c
Correct.

**Comment by guymcswain** [comment]
Yes, but I only see a subset of the bits getting set.  Also, not sure if they are cleared when a 0 gets or'ed for that position.

**Comment by guymcswain** [comment]
To verify, this same code failed before you edited the single line in pigpio and is now working?

**Comment by smeerpate** [comment]
Hi Guy,
It seems so on my UsingGpioLib-branch.
I will verify that to be 100% sure and I will get back to you on this later this week.
Thanks for your help!
Frederic

**Comment by smeerpate** [comment]
Hi Guy,
I have checked this again, and as you suspected, the volatile statement doesn’t change a thing. I still don’t understand why I’m reading two different values in the control register.
My program here outputs this:
![78EB8564-82B5-4BE3-AB63-5895FF9C40D3](https://user-images.githubusercontent.com/51826342/103077240-b7417800-45cf-11eb-9261-daa96dd80184.jpeg)
This is what I read back when I’m using another program just after I read back 0x305 from CR using mmap:
![0F54A728-60A0-4E94-B527-954E21F6F946](https://user-images.githubusercontent.com/51826342/103077614-6716e580-45d0-11eb-96a0-e4daeec418aa.jpeg)
I finally moved over to using the serial port with termios instead because I ran out of time...
Best regards,
Frederic

**Comment by guymcswain** [comment]
Hi Frederic,

I'm going to write a test that will try to simulate the conditions in your environment.  To that end I have a few questions:
1. Which RPi model are you using?
2. Is the I2C master external and if so what is it?
3. OS and pigpio versions?

Thanks,
Guy

**Comment by smeerpate** [comment]
Hi Guy,

1. cat /proc/device-tree/model:
> Raspberry Pi 3 Model B Rev 1.2

2. External I2c master:
> Arduino micro (P/N A000053)

3a. cat /etc/os-release:
> PRETTY_NAME="Raspbian GNU/Linux 9 (stretch)"
> NAME="Raspbian` GNU/Linux"
> VERSION_ID="9"
> VERSION="9 (stretch)"
> ID=raspbian
> ID_LIKE=debian

3b. pigpio version:
> pigpio version 78

Does all above info makes sense?

I admire your commitment!

Best regards,
Frederic
